### ISSUE TITLE
Structural clipboard design

### ISSUE THREAD

**Issue by kchanqvq** [neomacs-project/neomacs#84]
Still not sure how to do this yet. The clipboard facility has to be able to copy content between different buffers (which expects different kind of structures).

**Comment by kchanqvq** [comment]
Alternatives considered:

1. Store (lists of) arbitrary Common Lisp objects in the clipboard
  Problem: This might loses information. E.g. if we store content in `lisp-mode` as S-exprs, line breaks are list.
2. Store DOM sub trees in the clipboard
  Problem: This seems to lose context-sensitive information as well. E.g. `web-copy-url` from `web-mode` and `copy-element` from `html-doc-mode` might both put an `<a>` element into the clipboard. In the former case, this looks like `<a href=url>title</a>`. In the latter case, it might be an `<a>` embeded in, say, a `<p>`. When pasting to `html-doc-mode`, this works as intended, but when pasting to `text-mode`, the `<a>` from `web-copy-url` should be pasted as an URL (its `href` attribute), but  the `<a>` from `html-doc-mode` need to be pasted as its text content (title). `text-mode` seems to have no way to tell.
  Workaround: add some kind of hint attribute to DOM nodes, e.g. `<a href=url to-text=title>title</a>`. However this just handles plain text. I feel like this is too hacky and probably not very extensible.
3. Store a key value map of `(mime-type, content)` in the clipboard
  This seems to be what every modern OS does. However I think this is too complicated and ideally every mode need to handle conversion from every mime-type (which probably isn't even a finite set).

@mmontone any ideas? :)

**Comment by mmontone** [comment]
I'm not sure either, this is not an easy problem :)
I would for sure have DOM for copy/paste that for same modes. So, at least don't lose any info when copy/paste occurs under same mode.
Then, the hard part, copy/paste across modes. I think I would attach info about the copied data and let each target mode (the mode on which the data is pasted) handle the copied data based on its attached type and info somehow; also perhaps the copy can attach multiple representations so that the target mode can pick. Something like that, but I'm not really sure.

**Comment by kchanqvq** [comment]
DOM with attached info looks like a good idea. In fact, maybe we can just reuse the `'presentation` attribute! (It is inspired by your Emacs PBUI, but I don't have multiple presentation types yet) I think that's already enough information. For the copy-url use case, just attach a string as presentation, then let text-mode always use the string if it see some node with a string presentation.

BTW, currently presentation is used for unprintable objects, context sensitives commands and completing-read. It seems that native CL type is sufficient so I don't have presentation type. How necessary do you think presentation types are?

**Comment by kchanqvq** [comment]
That's very helpful! Thanks for the pleasant discussion, as always :)

**Comment by kchanqvq** [comment]
BTW, I think the need for presentation type in Neomacs is reduced because the DOM itself contains structured data, and can serve the purpose of presentation types if needed.

**Comment by mmontone** [comment]
> That's very helpful! Thanks for the pleasant discussion, as always :)

Of course! Btw, I would like to have a chat with you at some point about some ideas I have in mind that could be interesting for the project. I know they would take time to implement, and I would like to get involved, but could be worth it, as neomacs is a special project with unique approach, and so it would be worth it to make it do great things. I'm talking about a generic "rich-text" mode, extensible and object oriented; also,  modes for org-mode and markdown or docbook, that kind of stuff ...

**Comment by kchanqvq** [comment]
Sure! Which platform would you like to chat? I'm currently locked out of IRC because my IP address and my account is stored in a machine that I don't have access...

**Comment by mmontone** [comment]
gitter.im ?

**Comment by kchanqvq** [comment]
Sure! I set up an account at @kchanqvq:gitter.im.

**Comment by attila-lendvai** [comment]
> then line breaks are lost

if one aims to write a strucutred editor, then i believe line breaks *should* be lost.

the various rendering parameters should be the setting of the observer.

**Comment by attila-lendvai** [comment]
the clipboard becomes a completely different beast when the domain is not a string of characters, but a complex structure.

i think it's worth forgetting everything we know about the clipboard, then come up with the new design, and only then try to make the behavior and the UI of the new model palatable for people coming from text editors.

i don't have time to get into all the details here, and i don't even remember them off the top of my head right now.

i don't know how much of the clipboard stuff is implemented in ProjecturEd. if you want to dvelve into its source then it's called `selection` there; `command`s can operate on a `selection`.
